### DISCUSSION TITLE
Matching up BJS features to Game elements

### DISCUSSION THREAD

**Discussion by jelster** [jelster/space-truckers#19] [General]
Since a goal of this project is to demonstrate various aspects of the BJS engine, it's important to try and ensure we cover at least the most important ones. This table is a list of BJS features along with notes, and rough priorities in some cases. 

We can try to create issues to track ones as we refine this

Let's see if we can get those creative juices flowing!

BabylonJS Feature or Concept | Space-Trucker Usage
-- | --
Animations | UI, Ambient   environment, cutscenes
Collisions | Trajectory   planning, space-highway hazards
Skeletons and   bones | Trucker model
Physics | Planning phase,   Obstacles and hazards
Particles | Rocket thrust,   solar storm, space debris. Particle Editor, NME (fragment shaders for   particles, if necessary - solar storm particles refraction effect) -   fireworks for high score screen - [Babylon.js Playground](https://playground.babylonjs.com/#D2FNK1#23)
SPS | ^^ - not   have-to-have, but worth looking at
Sprites | GUI, Scoreboard,   Sprite Editor
Audio | Music, sfx
GUI | 2D GUI + 3D GUI -   truck cabin
Behaviors | Lower priority
Node Material | Cell shading   post-process
PBR | Cargo pods, truck   models
Skybox | Space scenery
Morphing | Obstacle collision   damage? Floaty-space slime-ball that morphs shapes, amoeba blob uses morph   targets
Render Targets | VR, minimap
Dynamic Textures | Texture inspector?
Reflections | Cargo pod surface,   windshield glass
Procedural   Textures | Star field example  - can be made in NME -- Highlight of destination in planning   phase
Meshes, Skeletons,   Bones | Trucker, rig   models
Parametric Shapes | Trail mesh, launch   vector arrow - possible use: boost strip
Decals | Space-dirt,   space-bugs on windshield
Height maps | Low-priority,   potential use: asteroid surfaces? - do in NME! Texture inspector as part of   workflow; paint out a landing strip in texture inspector for truck stop
CSG (boolean   operations on meshes) | Low-priority;   potential use: road sign
Occlusion queries | Space-rig   headlights, space-highway lights, signage lighting
Offline mode | PWA, IndexedDB
Incremental   loading | Loading, Main menu   screens
Scene optimizer | TBD - mobile perf   optimization?
LOD | TBD - kind of a   given, so not necessary to cover as separate topic. Useful to incorporate   into related discussions, however. Adds "street cred" - AutoLOD
Offscreen canvas | TBD - low priority
Engine sharing | n/a
Fog | Space-highway
Shadow maps, CSM | Space-highway   scene - closer shadows get higher res
Post-Processes | Cell shader   (cartoon) effect, scene transitions - end-game fail state. Use NME to create
Reflection Probes | See PBR
Multi-views | Minimap, WebXR
Glow layer | Low priority
Cameras (VJ, WXR,   GP, etc) | Device input   management
Importers and   tooling | Development   workflow, NME
Instances | Similar to SPS   usage - multiple instances onscreen simultaneously, not cloned, not multiple   meshes, just one GL call

**Comment by eldinor** [comment]
| BabylonJS Feature or Concept       | Space-Trucker Usage                                                                                                                                                                                                                                              | First Comments                                                                                                                                                              |
| ---------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| Animations                         | UI, Ambient environment, cutscenes                                                                                                                                                                                                                               | First of all, camera animation.                                                                                                                                             |
| Collisions                         | Trajectory planning, space-highway hazards                                                                                                                                                                                                                       | Collisions or intersections (seems that in this case we’ll use intersections). Performance eater, be careful.                                                               |
| Skeletons and bones                | Trucker model                                                                                                                                                                                                                                                    | This is the very complex subject, and I am not sure that we need to skeletons/bones here.                                                                                   |
| Physics                            | Planning phase, Obstacles and hazards                                                                                                                                                                                                                            | There are 3 physics engines in Babylon: Ammo, Oimo and Cannon. Which one to choose and why? Physics callbacks, like registerOnPhysicsCollide                                |
| Particles                          | [Rocket thrust, solar storm, space debris. Particle Editor, NME (fragment shaders for particles, if necessary - solar storm particles refraction effect) - fireworks for high score screen - Babylon.js Playground](https://playground.babylonjs.com/#D2FNK1#23) | SPS also could go here, could be good for asteroid storm, for example.                                                                                                      |
| SPS                                | ^^ - not have-to-have, but worth looking at                                                                                                                                                                                                                      | See above.                                                                                                                                                                  |
| Sprites                            | GUI, Scoreboard, Sprite Editor                                                                                                                                                                                                                                   | Sprites have a lot of common with Texture Atlas (but I am not sure that we need it).                                                                                        |
| Audio                              | Music, sfx                                                                                                                                                                                                                                                       | Sound assets.                                                                                                                                                               |
| GUI                                | 2D GUI + 3D GUI - truck cabin                                                                                                                                                                                                                                    | We can also use somewhere plain HTML/CSS over the canvas (strangely, a lot of people don’t know about it, I don’t know why).                                                |
| Behaviors                          | Lower priority                                                                                                                                                                                                                                                   | There 2 types of behaviors: camera behavior and mesh behavior (for example, PointerDragBehavior). Not sure which we’ll use if any.                                          |
| Node Material                      | Cell shading post-process                                                                                                                                                                                                                                        | NME still has no guide, so I think we’ll need here short intro to GLSL.                                                                                                     |
| PBR                                | Cargo pods, truck models                                                                                                                                                                                                                                         | Short intro to physically based rendering. But we also can use simple standard materials if needed (for SPS, for example).                                                  |
| Skybox                             | Space scenery                                                                                                                                                                                                                                                    | Could be dynamic too (slowly rotating, for example).                                                                                                                        |
| Morphing                           | Obstacle collision damage? Floaty-space slime-ball that morphs shapes, amoeba blob uses morph targets                                                                                                                                                            | A lot of potential uses but quite complex topic.                                                                                                                            |
| Render Targets                     | VR, minimap                                                                                                                                                                                                                                                      |                                                                                                                                                                             |
| Dynamic Textures                   | Texture inspector?                                                                                                                                                                                                                                               | To show how to use dynamic 2D canvas inside 3D canvas.                                                                                                                      |
| Reflections                        | Cargo pod surface, windshield glass                                                                                                                                                                                                                              | As a part of PBR materials.                                                                                                                                                 |
| Procedural Textures                | Star field example - can be made in NME -- Highlight of destination in planning phase                                                                                                                                                                            |                                                                                                                                                                             |
| Meshes, Skeletons, Bones           | Trucker, rig models                                                                                                                                                                                                                                              | Skeletons, Bones – duplicate above. But we’ll need to explain somewhere the mesh structure (maybe together with Inspector). Mesh disposing and Garbage collector.           |
| Parametric Shapes                  | Trail mesh, launch vector arrow - possible use: boost strip                                                                                                                                                                                                      | Lines as well.                                                                                                                                                              |
| Decals                             | Space-dirt, space-bugs on windshield                                                                                                                                                                                                                             | For sure :)                                                                                                                                                                 |
| Height maps                        | Low-priority, potential use: asteroid surfaces? - do in NME! Texture inspector as part of workflow; paint out a landing strip in texture inspector for truck stop                                                                                                | Height maps are good for planes. For meshes like spheres/asteroids we can use displacement maps.                                                                            |
| CSG (boolean operations on meshes) | Low-priority; potential use: road sign                                                                                                                                                                                                                           |                                                                                                                                                                             |
| Occlusion queries                  | Space-rig headlights, space-highway lights, signage lighting                                                                                                                                                                                                     | There should be separate chapter about lighting/environment, which is very important.                                                                                       |
| Offline mode                       | PWA, IndexedDB                                                                                                                                                                                                                                                   | PWA as the future of web apps and how to build it. IndexedDB limitations and advantages.                                                                                    |
| Incremental loading                | Loading, Main menu screens                                                                                                                                                                                                                                       | Could sophisticate all things. Not sure that we need it.                                                                                                                    |
| Scene optimizer                    | TBD - mobile perf optimization?                                                                                                                                                                                                                                  | We should optimize as much as possible. How to measure performance (a lot of people don’t know about it).                                                                   |
| LOD                                | TBD - kind of a given, so not necessary to cover as separate topic. Useful to incorporate into related discussions, however. Adds "street cred" - AutoLOD                                                                                                        | Should be used for better rendering performance.                                                                                                                            |
| Offscreen canvas                   | TBD - low priority                                                                                                                                                                                                                                               |                                                                                                                                                                             |
| Engine sharing                     | n/a                                                                                                                                                                                                                                                              |                                                                                                                                                                             |
| Fog                                | Space-highway                                                                                                                                                                                                                                                    |                                                                                                                                                                             |
| Shadow maps, CSM                   | Space-highway scene - closer shadows get higher res                                                                                                                                                                                                              | All shadows are perf eaters.                                                                                                                                                |
| Post-Processes                     | Cell shader (cartoon) effect, scene transitions - end-game fail state. Use NME to create                                                                                                                                                                         | Use with care for performance.                                                                                                                                              |
| Reflection Probes                  | See PBR                                                                                                                                                                                                                                                          | Not sure that we need them. Perf expensive.                                                                                                                                 |
| Multi-views                        | Minimap, WebXR                                                                                                                                                                                                                                                   | Second camera for minimap?                                                                                                                                                  |
| Glow layer                         | Low priority                                                                                                                                                                                                                                                     | I think we’ll use a lot of emissive textures so there will be no problem to show glow layer as well. Glow layer allows to simulate lights, which are performance expensive. |
| Cameras (VJ, WXR, GP, etc)         | Device input management                                                                                                                                                                                                                                          | Camera tuning as one of the most important tasks.                                                                                                                           |
| Importers and tooling              | Development workflow, NME                                                                                                                                                                                                                                        | DCC, image/texture editors (just to mention them since they don’t relate to Babylon directly).                                                                              |
| Instances                          | Similar to SPS usage - multiple instances onscreen simultaneously, not cloned, not multiple meshes, just one GL call                                                                                                                                             | Difference between intances and clones. Good practice – instance as much as possible.                                                                                       |

**Reply by jelster** [reply]
Great analysis on that, @eldinor! I’ll incorporate this into the OP. 

Couple things I might clarify:

### Re: using html/css and such

you’re right about it being a powerful and often overlooked approach. I have so far opted against using such elements for the time being because I want to be able to build this using Babylon Native (as much as can be expected to be supported for the time being). A consequence of Native is that we can’t depend on anything that uses the DOM or HTML/Canvas elements, so anything done in those would also need to be built using non-HTML means. That’s not necessarily the worst thing in the world, but we will need to ensure we have that properly decoupled from the rest of the app code. 

### re: physics engine choices

good question. I’ve played around with all three, but I have to admit ignorance as to the quantitative relative strengths and weaknesses that distinguish them. Any preferences on your part?

### re: lighting and environment

The environmental refinement, including lighting and shadows, is definitely getting its own chapter in an upcoming section, but we have time - it’s scheduled for about 4-5 chapter from where we currently stand. We’ll go through creating a `.env` texture file from a prefiltered DDS, do IBL with PBR step, etc.

**Comment by eldinor** [comment]
### Re: using html/css and such
At least it worth to mention that it is possible.

### physics engine choices
All of them are worse or better or neutral for specific cases/apps. It is not easy question.
Also worth to mention that it is possible to use your own physics engine.

**Comment by jelster** [comment]
Moved this into general since it's not a q and a type of post
